### PR TITLE
UPSTREAM: <carry>: Add process overlap detection event to watch-termination

### PR DESCRIPTION
**Pull request by mfojtik**

### CODE REVIEW

File: cmd/watch-termination/main.go [hunk 1]
```diff
@@ -79,6 +80,33 @@ func run() int {
 		}
 	}
 
+	if processOverlapDetectionFile != nil && len(*processOverlapDetectionFile) > 0 {
+		_, err := ioutil.ReadFile(*processOverlapDetectionFile)
+		if err != nil && !os.IsNotExist(err) {
+			klog.Errorf("failed to read lock timeout file: %v", err)
```
Comment (sttts) [inline cmd/watch-termination/main.go:86]: why timeout?

File: cmd/watch-termination/main.go [hunk 2]
```diff
@@ -79,6 +80,33 @@ func run() int {
...
+			return 1
+		}
+		if err == nil {
+			if client != nil {
```
Comment (sttts) [inline cmd/watch-termination/main.go:90]: how can this be?

File: cmd/watch-termination/main.go [hunk 3]
```diff
@@ -79,6 +80,33 @@ func run() int {
 		}
 	}
 
+	if processOverlapDetectionFile != nil && len(*processOverlapDetectionFile) > 0 {
+		_, err := ioutil.ReadFile(*processOverlapDetectionFile)
+		if err != nil && !os.IsNotExist(err) {
+			klog.Errorf("failed to read lock timeout file: %v", err)
+			return 1
+		}
+		if err == nil {
```
Comment (sttts) [inline cmd/watch-termination/main.go:89]: return early

File: cmd/watch-termination/main.go [hunk 4]
```diff
@@ -79,6 +80,33 @@ func run() int {
...
+			if client != nil {
+				ref, err := eventReference()
+				if err != nil {
+					klog.Errorf("failed to get event target: %v", err)
+					return 1
+				}
+				go wait.PollUntil(5*time.Second, func() (bool, error) {
+					if err := eventf(client.CoreV1().Events(ref.Namespace), *ref, corev1.EventTypeWarning, "TerminationProcessOverlapDetected", "The kube-apiserver setup fallback to slow run because of kubelet or container runtime error."); err != nil {
```
Comment (sttts) [inline cmd/watch-termination/main.go:97]: ERRFAILEDPARSINGSENTENCE

File: cmd/watch-termination/main.go [hunk 5]
```diff
@@ -79,6 +80,33 @@ func run() int {
...
+			return 1
+		}
+		if err == nil {
+			if client != nil {
+				ref, err := eventReference()
+				if err != nil {
+					klog.Errorf("failed to get event target: %v", err)
+					return 1
+				}
+				go wait.PollUntil(5*time.Second, func() (bool, error) {
```
Comment (sttts) [inline cmd/watch-termination/main.go:96]: we have the retry package

File: cmd/watch-termination/main.go [hunk 6]
```diff
@@ -14,6 +14,8 @@ import (
 	"syscall"
 	"time"
 
+	"k8s.io/kubernetes/test/utils"
+
```
Comment (sttts) [inline cmd/watch-termination/main.go:18]: test utils?

File: cmd/watch-termination/main.go [hunk 7]
```diff
@@ -31,6 +32,7 @@ func main() {
 func run() int {
 	terminationLog := flag.String("termination-log-file", "", "Write logs after SIGTERM to this file (in addition to stderr)")
 	terminationLock := flag.String("termination-touch-file", "", "Touch this file on SIGTERM and delete on termination")
+	processOverlapDetectionFile := flag.String("process-overlap-detection-file", "", "Presence of this file means the previous API servers was not properly terminated by kubelet and we had to fallback to slow run")
```
Comment (sttts) [inline cmd/watch-termination/main.go:35]: "and we had to fallback to slow run" sounds strange.

Use the wording of the event.

**Comment by mfojtik** [comment]
/hold

for rebase landing

**Comment by soltysh** [comment]
Included in #862 and squashed to `UPSTREAM: <carry>: Add OpenShift tooling, images, configs and docs commit.`
/close

**Pull request closed without merging**